### PR TITLE
Switch to transforming on a per character basis for non-valid utf8 messages

### PR DESCRIPTION
**Pull request by starbelly**
This PR adjusts the work done in #21. An edge case can happen with blanket coercing a non-valid utf8 string into a valid utf8 string as mentioned here : https://www.erlang.org/doc/apps/stdlib/unicode_usage.html#double-utf-8-encoding

See the test case added for an example. Unfortunately, this means when the string in question is not valid utf8 the safest way to perform the conversion is char by char. This doesn't seem to have a huge performance penalty.

### CODE REVIEW

File: lib/hl7/message.ex [hunk 1]
```diff
@@ -508,18 +508,15 @@ defmodule HL7.Message do
...
   end
 
-  defp validate_text(raw_text, options) do
-    encoded_text =
-      if options[:accept_latin1] == true do
-        :unicode.characters_to_binary(raw_text, :latin1)
-      else
-        raw_text
-      end
+  defp validate_text(raw_text, %{accept_latin1: true}) do
```
Comment (starbelly) [inline lib/hl7/message.ex:511]: I moved this into its own clause since there should be no need to validate after a transform (if one happens).

File: lib/hl7/message.ex [hunk 2]
```diff
@@ -529,6 +526,30 @@ defmodule HL7.Message do
     end
   end
 
+  def maybe_transcode(binary) do
+    case String.valid?(binary) do
```
Comment (starbelly) [inline lib/hl7/message.ex:530]: `String.valid?/1` seems to be slightly faster than `:unicode.characters_to_binary/1`,  it may be it's still best to stick with the unicode module for this.

File: lib/hl7/message.ex [hunk 3]
```diff
@@ -529,6 +526,30 @@ defmodule HL7.Message do
...
+      true ->
+        binary
+
+      false ->
+        transcode(binary)
+    end
+  end
+
+  def transcode(binary) do
```
Comment (starbelly) [inline lib/hl7/message.ex:539]: I found this was the most performant method to accumulate. If that's so, then it may invalidate the need to perform a copy if that option is passed in.

File: lib/hl7/message.ex [hunk 4]
```diff
@@ -529,6 +526,30 @@ defmodule HL7.Message do
...
+      false ->
+        transcode(binary)
+    end
+  end
+
+  defp transcode(binary) do
+    transcode(binary, [])
+  end
+
+  defp transcode(<<>>, acc), do: acc |> Enum.reverse() |> to_string()
```
Comment (starbelly) [inline lib/hl7/message.ex:543]: Between this version and basically how String.valid/1 is implemented, it may be best just to go with the String.valid/1 approach since it's simpler. No strong feelings.

File: lib/hl7/message.ex (hunk 2 above)
Comment (starbelly) [reply lib/hl7/message.ex:530]: I also don't think we should even check to see if the String is valid in the case accept_latin1 is true. We avoid iterating twice and if's invalid, we'll have to do the iteration anyway.

File: lib/hl7/message.ex (hunk 4 above)
Comment (josevalim) [reply lib/hl7/message.ex:543]: Please benchmark but I believe the Erlang VM can optimize the binary one really well.

Comment (starbelly) [reply lib/hl7/message.ex:543]: Your simpler implementation is faster about 14% percent faster. ❤️

File: lib/hl7/message.ex [hunk 5]
```diff
@@ -508,18 +508,15 @@ defmodule HL7.Message do
...
 
-  defp validate_text(raw_text, options) do
-    encoded_text =
-      if options[:accept_latin1] == true do
-        :unicode.characters_to_binary(raw_text, :latin1)
-      else
-        raw_text
-      end
+  defp validate_text(raw_text, %{accept_latin1: true}) do
+    {:ok, transcode(raw_text)}
```
Comment (starbelly) [inline lib/hl7/message.ex:512]: Per last change, just ripped out the validation for this case. It gets validated per the transform implicitly. I can not think of any performance reason why we would want to do essentially the same operation twice. It's not exactly twice (i.e., String.valid?/1 will bail early on), but for any binary where String.valid?/1 returns true, we'll pay the toll twice.

**Comment by DarkMarmot** [comment]
> This PR adjusts the work done in #21. An edge case can happen with blanket coercing a non-valid utf8 string into a valid utf8 string as mentioned here : https://www.erlang.org/doc/apps/stdlib/unicode_usage.html#double-utf-8-encoding
> 
> See the test case added for an example. Unfortunately, this means when the string in question is not valid utf8 the safest way to perform the conversion is char by char. This doesn't seem to have a huge performance penalty.

The nice thing about running `String.valid?` on the whole puppy is that it lets us generate quick `InvalidMessage` for pure junk I think? 

I think we might want to keep the latin1 naming as well. If i understand correctly, we can generate incorrect characters for latin9, latin15, etc if we encounter those encodings?

And running benchmarks, this looks like it would drop our performance by at least 2x. I get 55k ips down to 26k ips on my laptop when I toggle the setting. I don't know if there's a good answer here, but it is a pretty heavy hit for us :(

**Comment by starbelly** [comment]
> The nice thing about running `String.valid?` on the whole puppy is that it lets us generate quick `InvalidMessage` for pure junk I think?

Oh you mean in the case the string is hopeless. Maybe. 🤔 

> I think we might want to keep the latin1 naming as well. If i understand correctly, we can generate incorrect characters for latin9, latin15, etc if we encounter those encodings?

That makes sense. I'm happy to change it back, I figured it made less sense after the change I made.

> 
> And running benchmarks, this looks like it would drop our performance by at least 2x. I get 55k ips down to 26k ips on my laptop when I toggle the setting. I don't know if there's a good answer here, but it is a pretty heavy hit for us :(

huh, so here's what I get : 

## Before

```
Name                 ips        average  deviation         median         99th %
raw             164.96 K        6.06 μs   ±130.77%        5.83 μs        6.79 μs
new              84.25 K       11.87 μs    ±20.85%       11.25 μs       20.42 μs
new-copy         78.55 K       12.73 μs    ±19.33%       12.13 μs       19.92 μs
round-trip       40.52 K       24.68 μs    ±14.16%       23.50 μs       37.35 μs
new-alt          24.13 K       41.44 μs     ±6.74%       40.67 μs       46.33 μs

Comparison:
raw             164.96 K
new              84.25 K - 1.96x slower +5.81 μs
new-copy         78.55 K - 2.10x slower +6.67 μs
round-trip       40.52 K - 4.07x slower +18.62 μs
new-alt          24.13 K - 6.84x slower +35.37 μs
```

## After

```
Name                 ips        average  deviation         median         99th %
raw             162.96 K        6.14 μs   ±112.50%        5.88 μs        8.21 μs
new              87.54 K       11.42 μs    ±49.08%       10.25 μs       43.25 μs
new-copy         78.99 K       12.66 μs    ±18.22%       12.08 μs       18.75 μs
round-trip       40.49 K       24.70 μs    ±15.44%       23.25 μs       37.54 μs
new-alt          24.36 K       41.05 μs     ±6.00%       40.21 μs       50.75 μs

Comparison:
raw             162.96 K
new              87.54 K - 1.86x slower +5.29 μs
new-copy         78.99 K - 2.06x slower +6.52 μs
round-trip       40.49 K - 4.03x slower +18.56 μs
new-alt          24.36 K - 6.69x slower +34.92 μs
```

Maybe try running the benchmark again? I've odd results with Benchee (i.e., yesterday with transcode functions)...

**Comment by starbelly** [comment]
> > The nice thing about running `String.valid?` on the whole puppy is that it lets us generate quick `InvalidMessage` for pure junk I think?
> 
> Oh you mean in the case the string is hopeless. Maybe. 🤔

@DarkMarmot What we could do is bail early in the transcode function if we find a non-utf8 character. The main thing was avoiding potentially iterating twice (especially since transcode does exactly what String.valid?/1 does, accumulation and transcoding aside that is).

File: lib/hl7/message.ex [hunk 6]
```diff
@@ -509,26 +509,50 @@ defmodule HL7.Message do
   end
 
   defp validate_text(raw_text, options) do
-    encoded_text =
-      if options[:accept_latin1] == true do
-        :unicode.characters_to_binary(raw_text, :latin1)
+    if options[:accept_latin1] do
```
Comment (starbelly) [inline lib/hl7/message.ex:512]: I've adjusted this such that it can possibly bail early, but it will only do so in case validate is true and accept_latin1 is false, and since we only transcode is accept_latin1 is true, it doesn't feel right (i.e., there's an execution path we'll never go down). 

Perhaps we need a case where accept_latin1 is true, we transcode but run into an error doing so, however I can not think of a case :)

**Comment by camilleryr** [comment]
Benchmarks 

on main
```
Name                 ips        average  deviation         median         99th %
raw              82.33 K       12.15 μs    ±57.62%       10.99 μs       40.99 μs
new              51.71 K       19.34 μs    ±39.79%       16.99 μs       54.99 μs
new-copy         44.74 K       22.35 μs    ±41.68%       18.99 μs       59.99 μs
round-trip       20.87 K       47.91 μs    ±33.31%       44.99 μs      122.99 μs
new-alt          14.42 K       69.34 μs    ±29.13%       62.99 μs      164.99 μs
```

feature branch
```
Name                 ips        average  deviation         median         99th %
raw              75.82 K       13.19 μs    ±70.39%          11 μs          51 μs
new-copy         43.00 K       23.26 μs    ±50.99%          19 μs          70 μs
new              19.18 K       52.14 μs    ±30.97%          48 μs         120 μs
round-trip       17.65 K       56.67 μs    ±33.95%          50 μs         139 μs
new-alt          12.58 K       79.46 μs    ±30.40%          70 μs         178 μs
```

**Comment by starbelly** [comment]
> Benchmarks
> 
> on main
> 
> ```
> Name                 ips        average  deviation         median         99th %
> raw              82.33 K       12.15 μs    ±57.62%       10.99 μs       40.99 μs
> new              51.71 K       19.34 μs    ±39.79%       16.99 μs       54.99 μs
> new-copy         44.74 K       22.35 μs    ±41.68%       18.99 μs       59.99 μs
> round-trip       20.87 K       47.91 μs    ±33.31%       44.99 μs      122.99 μs
> new-alt          14.42 K       69.34 μs    ±29.13%       62.99 μs      164.99 μs
> ```
> 
> feature branch
> 
> ```
> Name                 ips        average  deviation         median         99th %
> raw              75.82 K       13.19 μs    ±70.39%          11 μs          51 μs
> new-copy         43.00 K       23.26 μs    ±50.99%          19 μs          70 μs
> new              19.18 K       52.14 μs    ±30.97%          48 μs         120 μs
> round-trip       17.65 K       56.67 μs    ±33.95%          50 μs         139 μs
> new-alt          12.58 K       79.46 μs    ±30.40%          70 μs         178 μs
> ```

That's a significant difference (to me)... unfortunately, I don't think we have a method that will be any faster and more importantly still safe.

**Comment by starbelly** [comment]
After a lot of thought we have decided to not support this as an option, instead users of this lib should perform encoding conversion ahead of passing a payload into HL7. We will keep the validation, but open up another PR to remove the accept_latin1 option. 

Cheers.
